File: backend/app/schemas/agent_action.py
"""Agent action schemas - the heart of the governance workflow."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.core.enums import ActionDecision, ActionStatus


class AgentActionCreate(BaseModel):
    agent_id: uuid.UUID
    resource: str = Field(..., min_length=1, max_length=100)
    action: str = Field(..., min_length=1, max_length=100)
    input_payload: dict[str, Any] = Field(default_factory=dict)


class AgentActionDecisionResponse(BaseModel):
    """Compact response returned by ``POST /agent-actions`` (matches the spec)."""

    agent_action_id: uuid.UUID
    decision: ActionDecision
    risk_score: int
    decision_reason: str
    status: ActionStatus
    approval_id: uuid.UUID | None = None


class AgentActionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID
    agent_id: uuid.UUID
    resource: str
    action: str
    input_payload: dict[str, Any]
    output_payload: dict[str, Any] | None
    risk_score: int
    decision: ActionDecision
    decision_reason: str
    status: ActionStatus
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/agentAction.ts
import type { ID, ISODateString, JsonObject } from './common'

export type Decision = 'ALLOW' | 'BLOCK' | 'PENDING_APPROVAL'

export interface AgentAction {
  id: ID
  agent_id: ID
  resource: string
  action: string
  input_payload: JsonObject
  decision: Decision
  risk_score: number
  decision_reason: string
  status: string
  matched_policy?: string | null
  created_at: ISODateString
}
